libcloudforensics/providers/kubernetes/base.py

```diff
@@ -0,0 +1,273 @@
+      name (str): The name of this resource.
+    """
+    super().__init__(api_client)
+    self.name = name
+
+  @abc.abstractmethod
+  def Read(self) -> object:
+    """Returns the resulting read operation for the resource.
+
+    For example, a Node resource would call CoreV1Api.read_node, a Pod
+    resource would call CoreV1Api.read_namespaced_pod. The return values
+    of these read calls do not share a base class, hence the object return
+    type.
+
+    Returns:
+      object: The result of this resource's matching read operation.
+    """
+
+
+class K8sCluster(K8sClient):
+  """Class representing a Kubernetes cluster."""
+
+  def __init__(self, api_client: client.ApiClient) -> None:
+    """Creates a K8sCluster object, checking the API client's authorization.
+
+    This constructor calls an authorization check on the api_client, to see
+    whether it is authorized to do all operations on the cluster. The equivalent
+    check for kubectl would be ``kubectl auth can-i '*' '*' --all-namespaces``.
+
+    Args:
+      api_client (client.ApiClient): The API client to the Kubernetes cluster.
+
+    Raises:
+      errors.CredentialsConfigurationError: If this cluster's API client is not
+        authorized for all operations on all resources in all namespaces of this
+        cluster.
+    """
+    super().__init__(api_client)
+    self.__AuthorizationCheck()
+
+  def ListPods(self, namespace: Optional[str] = None) -> List['K8sPod']:
+    """Lists the pods of this cluster, possibly filtering for a namespace.
+
+    Args:
+      namespace (str): Optional. The namespace in which to list the pods.
+
+    Returns:
+      List[K8sPod]: The list of pods for the namespace, or in all namespaces
+        if none is specified.
+    """
+    api = self._Api(client.CoreV1Api)
+
+    # Collect pods
+    if namespace is not None:
+      pods = api.list_namespaced_pod(namespace)
+    else:
+      pods = api.list_pod_for_all_namespaces()
+
+    # Convert to node objects
+    return [K8sPod(self._api_client, pod.metadata.name, pod.metadata.namespace)
+            for pod in pods.items]
+
+  def ListNodes(self) -> List['K8sNode']:
+    """Lists the nodes of this cluster.
+
+    Returns:
+      List[K8sNode]: The list of nodes in this cluster.
+    """
+    api = self._Api(client.CoreV1Api)
+
+    # Collect pods
+    nodes = api.list_node()
+
+    # Convert to node objects
+    return [K8sNode(self._api_client, node.metadata.name)
+            for node in nodes.items]
+
+  def __AuthorizationCheck(self) -> None:
+    """Checks the authorization of this cluster's API client.
+
+    Performs a check as per ``kubectl auth can-i '*' '*' --all-namespaces``.
+
+    Raises:
+      errors.CredentialsConfigurationError: If this cluster's API client is not
+        authorized for all operations on all resources in all namespaces of this
+        cluster.
+    """
+    api = self._Api(client.AuthorizationV1Api)
+    response = api.create_self_subject_access_review(
```

Comment:
isn-t it better to use create_self_subject_rules_review  (similar to kubectl auth can-i --list) and get a list of all actions, then error if no action can be performed or if the minimum required set of actions is not returned?
you can also do create_self_subject_access_review on all actions and  log a warning if not possible. But raising an error and blocking if no full access to all namespaces granted may lead to requiring more permissions than needed (I am imagining an enumeration function won't need a full set of actions on all resources)

Reply:
I think that rather than checking minimum set of requirements, maybe we should just let the API calls fail instead. This minimum set might be constantly changing and could introduce a lot of boilerplate code IMO.

I see two approaches:
- A wrapper function to all API calls, which takes care of checking whether or not the call failed due to not enough permissions
- Just logging a warning about not having all permissions, so some calls may fail (as you mentioned)

Reply:
logging a warning that some permissions are missing is good. If no enough permissions exceptions will be surfaced and calls will fail anyhow.

Later-on if there are some high level functions that require chaining low-level actions and some of these actions take long time to finish, its worth checking permissions in advance for all actions to avoid failing after waiting too long. For now a warning is good enough.